### PR TITLE
Allow inference on multiple images at once

### PR DESCRIPTION
**Pull request by victor-psiori**
Ticket: [https://psiori.atlassian.net/browse/WOL-159](https://psiori.atlassian.net/browse/WOL-159)

### CODE REVIEW

File: modify_model.py [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+        img_tensor = tf.convert_to_tensor(original_image, dtype=tf.float32)
+
+        # (hxwx3) -> (1xhxwx3)
+        img_tensor = tf.expand_dims(img_tensor, axis=0)
+        output_tensors.append(img_tensor)
+    output_tensors = tf.concat(output_tensors, axis=0)
+    return output_tensors
+
+
+def create_yolo_model(args):
```
Comment (ahoereth) [inline modify_model.py:159]: confused this is needed, feels like it should live somewhere else for deduplication.

File: modify_model.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+        if i == 0:
+            output_tensors = decode(
+                fm,
+                args.input_size // 8,
+                NUM_CLASS,
+                STRIDES,
+                ANCHORS,
+                i,
+                XYSCALE=XYSCALE,
+                FRAMEWORK="onnxtrt",
```
Comment (ahoereth) [inline modify_model.py:178]: I don't think FRAMEWORK="onnxtrt" does anything? Results in "tf" if I look into that method. Misleading to pass noop argument values.

File: modify_model.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+            )
+        bbox_tensors.append(output_tensors[0])
+        prob_tensors.append(output_tensors[1])
+    # -> [B, num_detections, 4]
+    pred_bbox = tf.concat(bbox_tensors, axis=1)
+    # -> [B, num_detections, 1]
+    pred_prob = tf.concat(prob_tensors, axis=1)
+
+    def add_onnxtrt_support():
```
Comment (ahoereth) [inline modify_model.py:209]: Should be deduplicated somewhere. Pass batch_size as argument and then it should also work for batch_size==1 without any changes, allowing it to be used in other scripts

**Review by ahoereth — COMMENTED** [review]
Did not review most of the formatting changes, if there are relevant changes in there please let me know. Reformatting a whole codebase without any code changes normally is better done in its own pr such that one does not need to search for proper changes.

Main concern with the code is that the codebase is getting worse by the day. I try to avoid to be an active contributor, so you are one of the main persons in charge to try to make it great. Deduplicating and commenting code is a good start on that front!

File: modify_model.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+            all_classes.append(curr_classes)
+            all_valid_detections.append(curr_valid_detections)
+
+        model = tf.keras.Model(
+            input_layer,
+            [
+                tf.stack(all_boxes, axis=0),  # -> [B, 50, 4]
+                tf.stack(all_scores, axis=0),  # -> [B, 50]
+                tf.stack(all_classes, axis=0),  # -> [B, 50]
+                all_valid_detections,  # [50, 50, 50, 50]
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py:252]: changed accordingly.

File: modify_model.py [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        help="Image directory.",
+    )
+    parser.add_argument(
+        "-m",
+        "--model",
+        type=str,
+        default="yolov4",
+        choices=["yolov4", "yolov3"],
+        help="Yolo model version",
+    )
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py:304]: removed `--model` cmd line argument. Only yolo4 supported.

File: modify_model.py [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+        type=float,
+        default=0.0,
+        help="score threshold",
+    )
+    parser.add_argument(
+        "--test_enabled",
+        dest="test_enabled",
+        action="store_true",
+        help="whether or not to run inference on batch of images after converting the model",
+    )
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py:328]: Inference on images done only if `image_dir` provided.

File: modify_model.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+        default=416,
+        help="input image size of trained model",
+    )
+    parser.add_argument(
+        "-i",
+        "--image_dir",
+        type=str,
+        default="./data/dataset/test_images",
+        help="Image directory.",
+    )
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py:296]: `image_dir` defaults to empty disabling tests.

File: modify_model.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+        type=str,
+        help="path to trained frozen model",
+    )
+    parser.add_argument(
+        "-o",
+        "--output_path",
+        default="./data/final_model",
+        type=str,
+        help="path to save tf trained model",
+    )
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py:275]: corrected accordingly.

File: modify_model.py [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+import matplotlib.pyplot as plt
+import glob
+
+
+# """
+# Function does following:
+#     - loads input yolo model trained with batch_size B = 1.
+#     - creates a tf keras model with inputs B = {2, 3, 4} and outputs:
+#       [boxes, scores, classes, valid_detections] corresponding to B > 1
+# """
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py]: corrected docstring with usage example.

File: modify_model.py (hunk 2 above)
Comment (victor-psiori) [reply modify_model.py:182]: Thanks for suggestion; didn't notice it then. removed noop argument `FRAMEWORK `.

File: modify_model.py [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+    for i, (img, box, score, cl, det) in enumerate(
+        zip(imgs, boxes, scores, classes, valid_det)
+    ):
+        painted_img = draw_bbox(img, [box, score, cl, det]).astype(np.float32)
+        fig, axs = plt.subplots(1, 1, figsize=(10, 10))
+        axs.imshow(painted_img)
+        fig.savefig(f"output_{i}.png")
+
+
+def preprocess_image(img: np.ndarray = None, output_size: List[int] = [416, 416]):
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py:132]: unified with utils `preprocess` function.

File: modify_model.py [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+    return STRIDES, ANCHORS, NUM_CLASS, XYSCALE
+
+
+def draw_bbox(
+    image,
+    bboxes,
+    classes=utils.read_class_names(cfg.YOLO.CLASSES),
+    show_label=True,
+    score_threshold=0.70,
+):
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py:47]: unified and refactored to methods in utils.

File: modify_model.py [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+                bbox_thick // 2,
+                lineType=cv2.LINE_AA,
+            )
+    return image
+
+
+def draw_batch_bbox(
+    img_tensor: tf.Tensor = None,
+    pred_result: List[tf.Tensor] = None,
+) -> None:
```
[the remark this answers is not in the document]
Comment (victor-psiori) [reply modify_model.py:108]: moved to utils.

File: modify_model.py (hunk 1 above)
Comment (victor-psiori) [reply modify_model.py:149]: I would prefet to keep this in `modify_model.py` file since this does the major heavy lifting in the file.

Comment (ahoereth) [reply modify_model.py:149]: But it’s a copy of the same code in another file? Should maybe go to the yolo file?

File: modify_model.py (hunk 3 above)
Comment (victor-psiori) [reply modify_model.py]: Did so. Refactored `add_onnxtrt_support` to utils and replaced function calls accordingly.

File: modify_model.py (hunk 1 above)
Comment (victor-psiori) [reply modify_model.py:60]: that's right. I extracted the common parts to `core/yolo.py` to deduplicate. let me know if that works for time being.

**Comment by victor-psiori** [comment]
Thanks for the review comments. :+1: 
Refactoring and clean ups are done from my side. 
Will run trigger a mini `train.py` job to make sure nothing is broken in training side.

**Comment by ahoereth** [comment]
Can you trigger a proper tiny yolo training run with the same config used for our best model currently? Want to see how it performs because this is too slow with 80ms..

**Comment by victor-psiori** [comment]
seeing the time data for tiny network (21.06 ms against 24ms in normal yolo), I am not sure whether there would be significant improvement in latency. I will run a train-run once the reported bug is solved.

File: core/utils.py [hunk 13]
```diff
@@ -72,49 +76,110 @@ def load_weights(model, weights_file, model_name='yolov4', is_tiny=False):
     wf.close()
 
 
+def add_onnxtrt_support(
+    input_layer: tf.Tensor = None,
+    pred_bbox: tf.Tensor = None,
+    pred_prob: tf.Tensor = None,
+    cmd_args: Dict = None,
+) -> tf.keras.Model:
```
Comment (ahoereth) [inline core/utils.py:84]: Methods should have a docstring

File: core/utils.py [hunk 14]
```diff
@@ -72,49 +76,110 @@ def load_weights(model, weights_file, model_name='yolov4', is_tiny=False):
...
+    all_boxes = []
+    all_scores = []
+    all_classes = []
+    all_valid_detections = []
+    # tensorrt does not support CombinedNonMaxSuppression, therefore we only
+    # predict a single category.
+    # Only support single image inference
+    for i in range(cmd_args["batch_size"]):
```
Comment (ahoereth) [inline core/utils.py:92]: Why not just have explicit arguments? batch_size, input_size, category_index, iou_threshold, score_threshold?

File: core/utils.py [hunk 15]
```diff
@@ -165,15 +254,52 @@ def draw_bbox(image, bboxes, classes=read_class_names(cfg.YOLO.CLASSES), show_la
...
+                bbox_thick // 2,
+                lineType=cv2.LINE_AA,
+            )
     return image
 
+
+def draw_batch_bbox(
+    img_tensor: tf.Tensor = None,
+    pred_result: List[tf.Tensor] = None,
+) -> None:
```
Comment (ahoereth) [inline core/utils.py:280]: docstring

File: core/yolov4.py [hunk 16]
```diff
@@ -122,109 +136,254 @@ def YOLOv4(input_layer, NUM_CLASS):
...
+def decode(
+    conv_output,
+    output_size,
+    NUM_CLASS,
+    STRIDES,
+    ANCHORS,
+    i,
+    XYSCALE=[1, 1, 1],
+    FRAMEWORK="tf",
+):
```
Comment (ahoereth) [inline core/yolov4.py:285]: docstring

File: modify_model.py [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    return final_model
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--input_model_path",
+        required=True,
+        type=str,
+        help="path to trained frozen model",
```
Comment (ahoereth) [inline modify_model.py:93]:
```suggestion
        help="Path to yolo saved in SavedModel format.",
```

File: modify_model.py [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+    )
+    parser.add_argument(
+        "-o",
+        "--output_path",
+        default="",
+        type=str,
+        help="path to save tf trained model",
```
Comment (ahoereth) [inline modify_model.py:100]:
```suggestion
        help="Path to save modifed model to. Optional, autogenerated if not provided.",
```

File: modify_model.py [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+        type=int,
+        default=1,
+        help="required batch size for inference. Possible values = {1, 2, 3, 4}",
+    )
+    parser.add_argument(
+        "-s",
+        "--input_size",
+        type=int,
+        default=416,
+        help="input image size of trained model",
```
Comment (ahoereth) [inline modify_model.py:114]: I feel like we should be able to determine that from the provided model. No need to change now..

File: modify_model.py [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+    )
+    parser.add_argument(
+        "-i",
+        "--image_dir",
+        type=str,
+        default="",
+        help="Image directory.",
```
Comment (ahoereth) [inline modify_model.py:121]:
```suggestion
        help="Directory of images for testing the modified model against.",
```

File: modify_model.py [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+    print(f"\n\t transfering weights to toy model from input model... \n")
+    toy_model = loaded_model.__class__.from_config(custom_config, custom_objects={})
+    weights = [layer.get_weights() for layer in loaded_model.layers[1:]]
+    for layer, w in zip(toy_model.layers[1:], weights):
+        layer.set_weights(w)
+    print(f"\n\n\n\t transfering weights to toy model successful \n\n\n")
+
+    final_model = create_yolo_model(args)
+    print(f"\n\t transfering weights to final model from toy model... \n")
+    src_weights = [layer.get_weights() for layer in toy_model.layers]
```
Comment (ahoereth) [inline modify_model.py:168]: Is the toy_model needed? I feel like you can just use loaded_model.layers instead, without the need for the additional intermediate from_config model creation.

File: modify_model.py [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+    print(f"\n\n\n\t saving final_model ... \n\n\n")
+    if args.output_path:
+        final_model.save(args.output_path)
+    else:
+        output_path = os.path.join(
+            os.getcwd(),
+            args.input_model_path,
+            f"yolo_batchsize_{args.batch_size}",
+        )
+        final_model.save(output_path)
```
Comment (ahoereth) [inline modify_model.py:184]: Overly complex, only one final_model.save needed
```suggestion
    if not args.output_path:
        output_path = os.path.join(
            os.getcwd(),
            args.input_model_path,
            f"yolo_batchsize_{args.batch_size}",
        )
    final_model.save(output_path)
```

File: modify_model.py [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+
+    if args.image_dir:
+        # img_paths = [
+        #     "./data/dataset/test_images/2021-10-06T10_01_17_040Z.png",
+        #     "./data/dataset/test_images/2021-10-06T11_58_15_772Z.png",
+        #     "./data/dataset/test_images/2021-10-06T09_44_01_676Z.png",
+        #     "./data/dataset/test_images/2021-10-06T11_48_24_010Z.png",
+        # ]
+        img_paths = glob.glob(args.image_dir + "/*.png")
+        # TODO: improve to infer on mulitples of 4 images at a time.
```
Comment (ahoereth) [inline modify_model.py:195]:
```suggestion
        # TODO: improve to infer on multiples of 4 images at a time.
```

Comment (ahoereth) [reply modify_model.py:195]:
```suggestion
        # TODO: improve to infer on any number of images, matching batch_size
```

File: pure_trt.py [hunk 24]
```diff
@@ -218,33 +260,100 @@ def main():
...
+                out_boxes, out_scores, out_classes, num_boxes = pred_bbox
+                print(
+                    f"bound box shape: {np.shape(out_boxes)}, out_scores: {np.shape(out_scores)}"
+                )
+                print(f"out_classes: {np.shape(out_classes)}, num_boxes: {num_boxes}")
+
+                img = Image.fromarray(img.astype(np.uint8))
+                img = cv2.cvtColor(np.array(img), cv2.COLOR_BGR2RGB)
+                cv2.imwrite(results_dir + image, img)
+        elif batch_size == 4:
```
Comment (ahoereth) [inline pure_trt.py:302]: I don't understand why the code is so specialized for batch size 4 vs 1, loops also work when they only run a single step.

**Review by ahoereth — COMMENTED** [review]
Its okay to merge for me if you think the proposed changes are out of scope

File: core/utils.py (hunk 14 above)
Comment (victor-psiori) [reply core/utils.py]: Replaced dict args with explicit args.

File: modify_model.py (hunk 17 above)
Comment (victor-psiori) [reply modify_model.py:95]: right, it doesn't. corrected as per suggestion.

File: modify_model.py (hunk 20 above)
Comment (victor-psiori) [reply modify_model.py:123]: corrected as per suggestion.

File: modify_model.py (hunk 22 above)
Comment (victor-psiori) [reply modify_model.py:186]: corrected as per suggestion.

File: modify_model.py (hunk 23 above)
Comment (victor-psiori) [reply modify_model.py:197]: corrected as per suggestion.

File: modify_model.py (hunk 18 above)
Comment (victor-psiori) [reply modify_model.py:102]: corrected as per suggestion.

File: pure_trt.py (hunk 24 above)
Comment (victor-psiori) [reply pure_trt.py:302]: combined the if-else conditions for variable batch sizes via for loop.

File: core/utils.py (hunk 13 above)
Comment (victor-psiori) [reply core/utils.py]: added docstring.

File: core/utils.py (hunk 15 above)
Comment (victor-psiori) [reply core/utils.py:284]: added docstring.

File: modify_model.py (hunk 21 above)
Comment (victor-psiori) [reply modify_model.py]: Removed the `toy_model` thing.

File: core/utils.py [hunk 25]
```diff
@@ -72,49 +76,120 @@ def load_weights(model, weights_file, model_name='yolov4', is_tiny=False):
...
+def add_onnxtrt_support(
+    input_layer: tf.Tensor = None,
+    pred_bbox: tf.Tensor = None,
+    pred_prob: tf.Tensor = None,
+    batch_size: int = None,
+    input_size: int = 416,
+    category_index: int = 0,
+    iou: float = 0.5,
+    score: float = 0.0,
+) -> tf.keras.Model:
```
Comment (ahoereth) [inline core/utils.py:94]: Sorry man, this is not how python docstrings work. Please look at any method in https://github/psiori/psipy

Comment (victor-psiori) [reply core/utils.py:94]: the link throws "This site can’t be reached" error. Looking it up and modifying docstring shortly.

Comment (victor-psiori) [reply core/utils.py]: corrected docstrings of methods newly added.

**Review by ahoereth — APPROVED** [review]

**Pull request merged**